File: src/main/java/cc/hyperium/internal/addons/misc/AddonManifestParser.java
package cc.hyperium.internal.addons.misc;

import cc.hyperium.internal.addons.AddonManifest;
import com.google.common.io.Files;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.*;
import java.nio.charset.Charset;
import java.util.jar.JarFile;
import java.util.zip.ZipEntry;

/**
 * Parses an addon manifest.
 */
public class AddonManifestParser {
    private JsonObject json = null;
    private final Gson gson = new Gson();

    /**
     * Parses an addon manifest from a Jar file.
     *
     * @param jar The Jar file to analyze.
     * @throws IOException If something goes wrong.
     */
    public AddonManifestParser(JarFile jar) throws IOException {
        InputStream jarInputStream = null;
        try {
            ZipEntry entry = jar.getEntry("addon.json");

            File jsonFile = File.createTempFile("json", "tmp");
            jsonFile.deleteOnExit();

            jarInputStream = jar.getInputStream(entry);
            copyInputStream(jarInputStream, new FileOutputStream(jsonFile));

            String contents = Files.toString(jsonFile, Charset.defaultCharset());

            JsonParser parser = new JsonParser();
            JsonObject json = parser.parse(contents).getAsJsonObject();

            if (!json.has("version") && !json.has("name") && !json.has("mainClass")) {
                throw new IOException("Invalid addon manifest (Needs name, version and mainClass)");
            }

            this.json = json;
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (jarInputStream != null) {
                jarInputStream.close();
            }
            jar.close();
        }
    }

    /**
     * Parses an addon manifest from a string.
     *
     * @param contents The manifest to parse.
     */
    public AddonManifestParser(String contents) {
        JsonParser parser = new JsonParser();
        JsonObject json = parser.parse(contents).getAsJsonObject();
        if (!json.has("version") && !json.has("name") && !json.has("mainClass")) {
            try {
                throw new IOException("Invalid addon manifest (Needs name, version and mainClass)");
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        this.json = json;
    }

    public AddonManifest getAddonManifest() {
        return gson.fromJson(json, AddonManifest.class);
    }

    @Override
    public String toString() {
        return json.toString();
    }

    private void copyInputStream(InputStream input, OutputStream output) {
        copyLarge(input, output, new byte[1024 * 4]);
    }

    private void copyLarge(InputStream input, OutputStream outputStream, byte[] buffer) {
        int n;

        try {
            while ((n = input.read(buffer)) != -1) {
                outputStream.write(buffer, 0, n);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/cc/hyperium/internal/addons/misc/AddonManifestParser.kt
package cc.hyperium.internal.addons.misc

import cc.hyperium.internal.addons.AddonManifest
import com.google.common.io.Files
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.io.*
import java.nio.charset.Charset
import java.util.jar.JarFile

class AddonManifestParser {
    private var json: JsonObject? = null
    private val gson = Gson()

    constructor(jar: JarFile) {

        var jarInputStream: InputStream? = null
        try {
            val entry = jar.getEntry("addon.json")

            val jsonFile = File.createTempFile("json", "tmp")
            jsonFile.deleteOnExit()

            jarInputStream = jar.getInputStream(entry)
            val os = FileOutputStream(jsonFile)
            copyInputStream(jarInputStream, os)

            val contents = Files.toString(jsonFile, Charset.defaultCharset())

            val parser = JsonParser()
            val json = parser.parse(contents).asJsonObject

            if (!json.has("version") && !json.has("name") && !json.has("mainClass")) {
                throw Exception("Invalid addon manifest (Needs name, version and mainClass)")
            }
            this.json = json
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            if (jarInputStream != null) jarInputStream.close()
            jar.close()
        }
    }

    constructor(contents: String) {
        val parser = JsonParser()
        val json = parser.parse(contents).asJsonObject

        if (!json.has("version") && !json.has("name") && !json.has("mainClass")) {
            throw Exception("Invalid addon manifest (Must include name, version and mainClass)")
        }
        this.json = json
    }

    fun getAddonManifest(): AddonManifest {
        return gson.fromJson(json, AddonManifest::class.java)
    }

    override fun toString(): String {
        return json!!.toString()
    }

    private fun copyInputStream(input: InputStream, output: OutputStream): Int {
        val count = copyLarge(input, output, ByteArray(1024 * 4))
        return if (count > Integer.MAX_VALUE) {
            -1
        } else count.toInt()
    }

    @Throws(IOException::class)
    private fun copyLarge(input: InputStream, output: OutputStream, buffer: ByteArray): Long {
        var count: Long = 0
        var n = 0
        while ({ n = input.read(buffer); n }() != -1) {
            output.write(buffer, 0, n)
            count += n.toLong()
        }
        return count
    }
}
